### PR TITLE
Deposit_handler Contract

### PR DESCRIPTION
**Pull request by sparqet**
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Feature
- Testing

## What is the new behavior?

- Implemented deposit_handler contract

## Does this introduce a breaking change?

No

### CODE REVIEW

**Review by Sk8erboi84 — CHANGES_REQUESTED** [review]
Create the initialize function and call it in the constructor and move everything from the constructor to the initialize function

**Comment by Sk8erboi84** [comment]
Document the function

File: src/exchange/deposit_handler.cairo [hunk 1]
```diff
@@ -144,45 +149,108 @@ mod DepositHandler {
         fn create_deposit(
```
Comment (zarboq) [inline src/exchange/deposit_handler.cairo:149]: missing role checks, contract should import IRoleModule and RoleModule and call it's initializer, role_store should be removed from this contract storage.

Comment (zarboq) [reply src/exchange/deposit_handler.cairo:149]: Also missing global reentrancy

File: src/exchange/deposit_handler.cairo [hunk 2]
```diff
@@ -144,45 +149,108 @@ mod DepositHandler {
...
+                self.data_store.read(),
+                self.event_emitter.read(),
+                self.deposit_vault.read(),
+                account,
+                params
+            )
         }
 
-        fn cancel_deposit(ref self: ContractState, key: felt252) { // TODO
+        fn cancel_deposit(ref self: ContractState, key: felt252) {
```
Comment (zarboq) [inline src/exchange/deposit_handler.cairo:165]: same

File: src/exchange/deposit_handler.cairo [hunk 3]
```diff
@@ -144,45 +149,108 @@ mod DepositHandler {
...
+                0, //starting_gas
+                keys::user_initiated_cancel(),
+                array!['Cancel Deposit'] //TODO should be empty string
+            )
         }
 
-        fn execute_deposit(
-            ref self: ContractState, key: felt252, oracle_params: SetPricesParams
-        ) { // TODO
+        fn execute_deposit(ref self: ContractState, key: felt252, oracle_params: SetPricesParams) {
```
Comment (zarboq) [inline src/exchange/deposit_handler.cairo:190]: same

Comment (zarboq) [reply src/exchange/deposit_handler.cairo:190]: also oracle price

File: src/exchange/deposit_handler.cairo [hunk 4]
```diff
@@ -144,45 +149,108 @@ mod DepositHandler {
...
+                min_oracle_block_numbers,
+                max_oracle_block_numbers,
+                keeper,
+                starting_gas: 0 // TODO starting_gas
+            };
+
+            execute_deposit_utils::execute_deposit(params);
         }
     }
+/// TODO not needed
```
Comment (zarboq) [inline src/exchange/deposit_handler.cairo:240]:
```suggestion
/// TODO no try catch, we need to find alternative
```

File: src/exchange/deposit_handler.cairo [hunk 5]
```diff
@@ -144,45 +149,108 @@ mod DepositHandler {
...
-        ) { // TODO
+        fn execute_deposit(ref self: ContractState, key: felt252, oracle_params: SetPricesParams) {
+            let data_store = self.data_store.read();
+            // let starting_gas = gas_left();
+            let execution_gas = gas_utils::get_execution_gas(data_store, 0);
+
+            self.execute_deposit_keeper(key, oracle_params, get_caller_address())
         }
 
         fn simulate_execute_deposit(
```
Comment (zarboq) [inline src/exchange/deposit_handler.cairo:198]: same

File: src/exchange/deposit_handler.cairo [hunk 6]
```diff
@@ -144,45 +149,108 @@ mod DepositHandler {
...
             ref self: ContractState, key: felt252, params: SimulatePricesParams
-        ) { // TODO
+        ) {
+            let oracleParams = Default::default();
+
+            self.execute_deposit_keeper(key, oracleParams, get_caller_address())
         }
 
         fn execute_deposit_keeper(
```
Comment (zarboq) [inline src/exchange/deposit_handler.cairo:206]: same

**Review by zarboq — CHANGES_REQUESTED** [review]

**Review by zarboq — COMMENTED** [review]
small fix in execute_deposit_keeper

**Review by zarboq — APPROVED** [review]

**Review by zarboq — APPROVED** [review]

**Review by Sk8erboi84 — APPROVED** [review]

**Pull request merged**